File: find-k-closest-elements/find-k-closest-elements.cs
public class Solution {
    public IList<int> FindClosestElements(int[] arr, int k, int x) {
        List<int> result = new List<int>();
        if(arr==null || arr.Length==0)
        {
            return new List<int>();
        }
        
        int n = arr.Length;
        int low =0;
        int high = n-k;
       
        int temp = binarysearch(arr,k,x,low,high);
        for(int i=temp;i<temp+k;i++)
        {
            result.Add(arr[i]);
        }
        
        return result;
        
        
    }
    
    private int binarysearch(int[] arr,int k,int x,int low,int high)
    {
         while(low<=high)
        {
             int first = low + (high-low)/2;
             int last = first + k -1;
            if(first-1>=0 && (x-arr[first-1] <= arr[last]-x))
            {
                high = first-1;
            }
            else if (last+1 <arr.Length && (x-arr[first] > arr[last+1]-x))
            {
                low = first+1;
            }
            else
            {
                return first;
            }
        }
        return low;
    }
}

File: src/com/madhurima/FindKClosestElements.java
// for two pointer based solution
// Time Complexity : O(n -k) + O(K) i.e while loop to adjust left and right + for loop to go over final left to right

// for binary search based solution
// Time Complexity : O(Log(n-k)) to search start of window + O(k) to go over the window

// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes

package com.madhurima;

import java.util.ArrayList;
import java.util.List;

public class FindKClosestElements {

    //two pointer based solution, move pointers left and right until distance between them is k
    public List<Integer> findClosestElements(int[] arr, int k, int x) {
        if(arr == null || arr.length == 0){
            return new ArrayList<>();
        }

        int n = arr.length;
        int left = 0;
        int right = n-1;

        while(right - left + 1 > k){
            int distLeft = x - arr[left];
            int distRight = arr[right] - x;

            if(distLeft > distRight){
                left++;
            }else{
                right--;
            }
        }

        List<Integer> result = new ArrayList<>();
        for(int i = left; i <= right; i++){
            result.add(arr[i]);
        }

        return result;

    }


    //Binary search based solution
    public List<Integer> findClosestElementsBinarySearch(int[] arr, int k, int x) {
        if(arr == null || arr.length == 0){
            return new ArrayList<>();
        }

        int n = arr.length;
        int low = 0;
        int high = n-k; //as from this point we need k elements

        while(low < high){
            int mid = low + (high - low)/2; //to avoid integer overflow //mid is the new start
            int distStart = x - arr[mid];
            int distEnd = arr[mid+k] - x;

            if(distStart > distEnd){
                low = mid + 1;
            }else{
                high = mid;
            }

        }

        List<Integer> result = new ArrayList<>();
        for(int i = low; i < low + k; i++){
            result.add(arr[i]);
        }

        return result;

    }


}
